### ISSUE TITLE
"Invalid key length" error

### ISSUE THREAD

**Issue by Kammy6679** [AzureAD/rms-sdk-for-cpp#59]
Following this page, we cannot build the RMS SDK on Ubuntu 14.10, instead, we got the following error message:
![222](https://cloud.githubusercontent.com/assets/10860444/8908625/bc4c0202-34af-11e5-8e14-6b0b439836be.png)

Using QT Creator 5.3.2, we can build the RMS SDK. But we cannot encrypt the PDF files. Please use your sample to reproduce this issue:
1. Run your sample
2. Input chengzhang_lin@msopentechtest01.onmicrosoft.com   (Password: Foxit2015)
    ![333333](https://cloud.githubusercontent.com/assets/10860444/8908747/604f4e5e-34b0-11e5-9b03-02cd94575684.png)
3. Click “Encrypt PFILE(templates)”
4. Choose a template to encrypt the file
5. The sample will crash.

Please see the attached log files:

```
rmsauth_154743-0727.log:
15:47:43 INF: FileCache: path: /home/lcz/.ms-ad/token_cache.dat
15:47:46 INF: AuthenticationContext: AuthenticationContext
15:47:46 INF: Authenticator: detectAuthorityType
15:47:46 INF: Authenticator: isAdfsAuthority: '0'
15:47:46 INF: AuthenticationContext: acquireToken
15:47:46 INF: AuthenticationContext: acquireTokenCommonAsync
15:47:46 INF: AuthenticationContext: createWebAuthenticationDialog
15:47:46 INF: AcquireTokenHandlerBase: AcquireTokenHandlerBase
15:47:46 INF: AcquireTokenHandlerBase: createCallState
15:47:46 INF: AcquireTokenHandlerBase: === Token Acquisition started:
    Authority: https://sts.aadrm.com/_sts/oauth/
    Resource: api.aadrm.com
    ClientId: 4a63455a-cfa1-4ac6-bd2e-0d046cf1c3f7
    CacheType: FileCache (0 items)
    Authentication Target: 0

15:47:46 INF: AcquireTokenInteractiveHandler: AcquireTokenInteractiveHandler
15:47:46 INF: AcquireTokenHandlerBase: runAsync
15:47:46 INF: AcquireTokenHandlerBase: preRunAsync
15:47:46 INF: Authenticator: updateFromTemplateAsync
15:47:46 INF: AcquireTokenHandlerBase: validateAuthorityType
15:47:46 INF: AcquireTokenHandlerBase: notifyBeforeAccessCache
15:47:46 INF: FileCache: onBeforeAccess
15:47:46 INF: FileCache: readCache
15:47:46 INF: TokenCache: deserialize
15:47:46 INF: AuthenticationResult: deserialize
15:47:46 INF: TokenCacheKey: TokenCacheKey::getHashCode(): authority: https://sts.aadrm.com/_sts/oauth/; resource: api.aadrm.com; clientId: f3295a1c-bcfe-4430-80bf-1c0a5636e8e8 uniqueId: ; tokenSubjectType_: 0; hashcode: 1718433536864167329;
15:47:46 INF: TokenCache: Deserialized 1 items to token cache.
15:47:46 INF: TokenCache: loadFromCache
15:47:46 INF: TokenCache: No matching token was found in the cache
15:47:46 INF: AcquireTokenInteractiveHandler: preTokenRequest
15:47:46 INF: AcquireTokenHandlerBase: preTokenRequest
15:47:46 INF: AcquireTokenInteractiveHandler: acquireAuthorization
15:47:46 INF: AcquireTokenInteractiveHandler: sendAuthorizeRequest
15:47:46 INF: AcquireTokenInteractiveHandler: includeFormsAuthParams
15:47:46 INF: AcquireTokenInteractiveHandler: createAuthorizationUri
15:47:46 INF: AcquireTokenInteractiveHandler: createAuthorizationRequest
15:48:06 INF: OAuth2Response: parseAuthorizeResponse
15:48:06 INF: AcquireTokenInteractiveHandler: verifyAuthorizationResult
15:48:06 INF: AcquireTokenHandlerBase: sendTokenRequestAsync
15:48:06 INF: AcquireTokenHandlerBase: sendHttpMessageAsync
15:48:06 INF: HttpHelperQt: sendPostRequestAndDeserializeJsonResponseAsync
15:48:06 INF: HttpHelperQt: createRequest
15:48:06 INF: HttpHelperQt: jobPost
15:48:06 INF: HttpHelperQt: deserializeTokenResponse
15:48:06 INF: HttpHelperQt: logRequestHeaders
15:48:06 INF: HttpHelperQt: --> Headers:
15:48:06 INF: HttpHelperQt: User-Agent: x-client-DM
15:48:06 INF: HttpHelperQt: Content-Type: application/x-www-form-urlencoded
15:48:06 INF: HttpHelperQt: x-client-Ver: 0.1
15:48:06 INF: HttpHelperQt: x-client-SKU: RMSAuth
15:48:06 INF: HttpHelperQt: client-request-id: {3f823d9a-bc03-45bf-bd38-1d7b60c5e662}
15:48:06 INF: HttpHelperQt: return-client-request-id: true
15:48:06 INF: HttpHelperQt: request url:  https://sts.aadrm.com/_sts/oauth2/token
15:48:06 INF: HttpHelperQt: request body: &redirect_uri=https://client.test.app&code=hza0f5VdK3hmGFJFWs0F_WAI3a_b9POaogT7rIHOQlXPkjgHsLlXFwTS2BEgQLiocZ818v7ogEBGpHOit1nfyO2sFcmaadiOCuEUJNQry2AnK_Tig0bbx5XzywwFmhn4-7LXSVDeZ7r7cGn2-XIwI7oS8hbxtO6azIZNK-sBuxE0gQFfxqn_GLpwZLZJUeRcPxWf5d6hTkhOg69ZdzGNJjz_3rJDcQshvCAK-1dWWVAqJ0RfWq8v5HKZX6LFqbluaIwKH7_Bm32rVYdPq5PeCtt5ecKzVix67vROVerzD3U&grant_type=authorization_code&client_id=4a63455a-cfa1-4ac6-bd2e-0d046cf1c3f7&resource=api.aadrm.com
15:48:07 INF: HttpHelperQt: verifyCorrelationIdHeaderInReponse
15:48:07 INF: HttpHelperQt: deserializeTokenResponse
15:48:07 INF: OAuth2Response: parseTokenResponse
15:48:07 INF: OAuth2Response: curTimeUTC: 07:48:07 07.27.15; expiresIn: 36000
15:48:07 INF: OAuth2Response: tonken expiresOn: 17:48:07 07.27.15 (1438019287)
15:48:07 INF: OAuth2Response: parseIdToken
15:48:07 INF: AcquireTokenInteractiveHandler: postTokenRequest
15:48:07 INF: AcquireTokenHandlerBase: postTokenRequest
15:48:07 INF: Authenticator: updateTenantId
15:48:07 INF: TokenCache: storeToCache
15:48:07 INF: TokenCacheKey: TokenCacheKey::getHashCode(): authority: https://sts.aadrm.com/_sts/oauth/; resource: api.aadrm.com; clientId: 4a63455a-cfa1-4ac6-bd2e-0d046cf1c3f7 uniqueId: ; tokenSubjectType_: 0; hashcode: 9142296010660041720;
15:48:07 INF: TokenCache: An item was added to the cache
15:48:07 INF: TokenCacheKey: TokenCacheKey::getHashCode(): authority: https://sts.aadrm.com/_sts/oauth/; resource: api.aadrm.com; clientId: 4a63455a-cfa1-4ac6-bd2e-0d046cf1c3f7 uniqueId: ; tokenSubjectType_: 0; hashcode: 9142296010660041720;
15:48:07 INF: AcquireTokenHandlerBase: postRunAsync
15:48:07 INF: AcquireTokenHandlerBase: === Token Acquisition finished successfully. An access token was retuned:
    Access Token Hash: 訯堃誩⒍犈嬔坭妒>良?廼0珉+f
    Refresh Token Hash: 錘iq?Z<5?ew饨僃JR榢缱鮘n
    Expiration Time: 17:48:07 07.27.15 (1438019287)
    User Hash: nullptr

15:48:07 INF: AcquireTokenHandlerBase: notifyAfterAccessCache
15:48:07 INF: FileCache: onAfterAccess
15:48:07 INF: FileCache: writeCache
15:48:07 INF: TokenCache: serialize
15:48:07 INF: TokenCache: Serializing token cache with 2 items.
15:48:07 INF: AuthenticationResult: serialize
15:48:07 INF: AuthenticationResult: serialize
15:48:12 INF: AuthenticationContext: AuthenticationContext
15:48:12 INF: Authenticator: detectAuthorityType
15:48:12 INF: Authenticator: isAdfsAuthority: '0'
15:48:12 INF: AuthenticationContext: acquireToken
15:48:12 INF: AuthenticationContext: acquireTokenCommonAsync
15:48:12 INF: AuthenticationContext: createWebAuthenticationDialog
15:48:12 INF: AcquireTokenHandlerBase: AcquireTokenHandlerBase
15:48:12 INF: AcquireTokenHandlerBase: createCallState
15:48:12 INF: AcquireTokenHandlerBase: === Token Acquisition started:
    Authority: https://sts.aadrm.com/_sts/oauth/
    Resource: api.aadrm.com
    ClientId: 4a63455a-cfa1-4ac6-bd2e-0d046cf1c3f7
    CacheType: FileCache (2 items)
    Authentication Target: 0

15:48:12 INF: AcquireTokenInteractiveHandler: AcquireTokenInteractiveHandler
15:48:12 INF: AcquireTokenHandlerBase: runAsync
15:48:12 INF: AcquireTokenHandlerBase: preRunAsync
15:48:12 INF: Authenticator: updateFromTemplateAsync
15:48:12 INF: AcquireTokenHandlerBase: validateAuthorityType
15:48:12 INF: AcquireTokenHandlerBase: notifyBeforeAccessCache
15:48:12 INF: FileCache: onBeforeAccess
15:48:12 INF: FileCache: readCache
15:48:12 INF: TokenCache: deserialize
15:48:12 INF: AuthenticationResult: deserialize
15:48:12 INF: TokenCacheKey: TokenCacheKey::getHashCode(): authority: https://sts.aadrm.com/_sts/oauth/; resource: api.aadrm.com; clientId: f3295a1c-bcfe-4430-80bf-1c0a5636e8e8 uniqueId: ; tokenSubjectType_: 0; hashcode: 1718433536864167329;
15:48:12 INF: TokenCache: Deserialized 2 items to token cache.
15:48:12 INF: TokenCache: loadFromCache
15:48:12 INF: TokenCache: An item matching the requested resource was found in the cache
15:48:12 INF: TokenCache: Local time UTC: '07:48:12 07.27.15', token expiresOn: 17:48:07 07.27.15 (1438019287)
15:48:12 INF: TokenCache: A matching token was found in the cache
15:48:12 INF: AcquireTokenHandlerBase: postRunAsync
15:48:12 INF: AcquireTokenHandlerBase: === Token Acquisition finished successfully. An access token was retuned:
    Access Token Hash: 訯堃誩⒍犈嬔坭妒>良?廼0珉+f
    Refresh Token Hash: 錘iq?Z<5?ew饨僃JR榢缱鮘n
    Expiration Time: 17:48:07 07.27.15 (1438019287)
    User Hash: nullptr

15:48:12 INF: AcquireTokenHandlerBase: notifyAfterAccessCache
15:48:12 INF: FileCache: onAfterAccess
15:49:04 INF: AuthenticationContext: AuthenticationContext
15:49:04 INF: Authenticator: detectAuthorityType
15:49:04 INF: Authenticator: isAdfsAuthority: '0'
15:49:04 INF: AuthenticationContext: acquireToken
15:49:04 INF: AuthenticationContext: acquireTokenCommonAsync
15:49:04 INF: AuthenticationContext: createWebAuthenticationDialog
15:49:04 INF: AcquireTokenHandlerBase: AcquireTokenHandlerBase
15:49:04 INF: AcquireTokenHandlerBase: createCallState
15:49:04 INF: AcquireTokenHandlerBase: === Token Acquisition started:
    Authority: https://sts.aadrm.com/_sts/oauth/
    Resource: api.aadrm.com
    ClientId: 4a63455a-cfa1-4ac6-bd2e-0d046cf1c3f7
    CacheType: FileCache (2 items)
    Authentication Target: 0

15:49:04 INF: AcquireTokenInteractiveHandler: AcquireTokenInteractiveHandler
15:49:04 INF: AcquireTokenHandlerBase: runAsync
15:49:04 INF: AcquireTokenHandlerBase: preRunAsync
15:49:04 INF: Authenticator: updateFromTemplateAsync
15:49:04 INF: AcquireTokenHandlerBase: validateAuthorityType
15:49:04 INF: AcquireTokenHandlerBase: notifyBeforeAccessCache
15:49:04 INF: FileCache: onBeforeAccess
15:49:04 INF: FileCache: readCache
15:49:04 INF: TokenCache: deserialize
15:49:04 INF: AuthenticationResult: deserialize
15:49:04 INF: TokenCacheKey: TokenCacheKey::getHashCode(): authority: https://sts.aadrm.com/_sts/oauth/; resource: api.aadrm.com; clientId: f3295a1c-bcfe-4430-80bf-1c0a5636e8e8 uniqueId: ; tokenSubjectType_: 0; hashcode: 1718433536864167329;
15:49:04 INF: TokenCache: Deserialized 2 items to token cache.
15:49:04 INF: TokenCache: loadFromCache
15:49:04 INF: TokenCache: An item matching the requested resource was found in the cache
15:49:04 INF: TokenCache: Local time UTC: '07:49:04 07.27.15', token expiresOn: 17:48:07 07.27.15 (1438019287)
15:49:04 INF: TokenCache: A matching token was found in the cache
15:49:04 INF: AcquireTokenHandlerBase: postRunAsync
15:49:04 INF: AcquireTokenHandlerBase: === Token Acquisition finished successfully. An access token was retuned:
    Access Token Hash: 訯堃誩⒍犈嬔坭妒>良?廼0珉+f
    Refresh Token Hash: 錘iq?Z<5?ew饨僃JR榢缱鮘n
    Expiration Time: 17:48:07 07.27.15 (1438019287)
    User Hash: nullptr

15:49:04 INF: AcquireTokenHandlerBase: notifyAfterAccessCache
15:49:04 INF: FileCache: onAfterAccess


rms_log_154743-2707.log:
15:47:43 INF: +RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="END_USER_LICENSES_UR"
15:47:43 INF: -RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="END_USER_LICENSES_UR" returning 0 result(s)
15:47:43 INF: +RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="PUBLISHING_LICENSES_UR"
15:47:43 INF: -RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="PUBLISHING_LICENSES_UR" returning 0 result(s)
15:47:43 INF: +RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="TEMPLATES_UR"
15:47:43 INF: -RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="TEMPLATES_UR" returning 0 result(s)
15:47:43 INF: +RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="CLOUD_DIAGNOSTICS_UR"
15:47:43 INF: -RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="CLOUD_DIAGNOSTICS_UR" returning 0 result(s)
15:47:43 INF: +RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="PERFORMANCE_UR"
15:47:43 INF: -RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="PERFORMANCE_UR" returning 0 result(s)
15:47:43 INF: +RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="DNS_CLIENT_RESULT"
15:47:43 INF: -RestClientCache::Lookup: cacheName="msopentechtest01.onmicrosoft.com", tag="DNS_CLIENT_RESULT" returning 0 result(s)
15:47:44 INF: ==> HttpClientQt::GET https://api.aadrm.com/my/v1/servicediscovery?email=chengzhang_lin@msopentechtest01.onmicrosoft.com
15:47:46 INF: Response StatusCode: 401
15:47:46 ERR: error: Host requires authentication
15:48:07 INF: ==> HttpClientQt::GET https://api.aadrm.com/my/v1/servicediscovery?email=chengzhang_lin@msopentechtest01.onmicrosoft.com
15:48:11 INF: Response StatusCode: 200
15:48:11 INF: ==> HttpClientQt::GET https://api.aadrm.com/my/v1/templates
15:48:12 INF: Response StatusCode: 401
15:48:12 ERR: error: Host requires authentication
15:48:12 INF: ==> HttpClientQt::GET https://api.aadrm.com/my/v1/templates
15:48:14 INF: Response StatusCode: 200
15:48:58 INF: ==> HttpClientQt::GET https://api.aadrm.com/my/v1/publishinglicenses
15:49:04 INF: Response StatusCode: 401
15:49:04 ERR: error: Host requires authentication
15:49:04 INF: ==> HttpClientQt::POST https://api.aadrm.com/my/v1/publishinglicenses
15:49:07 INF: Response StatusCode: 200
```

NOTE:
If use QTStreamImpl to encrypt the file, you can get the following information by 

```
shared_future<int64_t> QTStreamImpl::WriteAsync(const uint8_t *cpbBuffer,
                                               int64_t        cbBuffer,
                                               int64_t        cbOffset,
                                               std::launch    launchType)
```

![44444](https://cloud.githubusercontent.com/assets/10860444/8908655/e0c99d74-34af-11e5-858e-ac2793489646.png)

**Comment by AgentRX** [comment]
Fixed in https://github.com/AzureAD/rms-sdk-for-cpp/pull/58
Now Azure uses AES 256 bit key.

**Comment by joshgav** [comment]
Fixed in 40c69e0c5d24cf5867bce96edf74f47587f10d46

**Comment by Kammy6679** [comment]
We still can run build_and_test.sh to build the RMS SDK, get the following error message:
![qq 20150728175020](https://cloud.githubusercontent.com/assets/10860444/8928454/2ce9bdfa-3551-11e5-99a5-bdca46ff9ecd.jpg)
